executor: desc table panic

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
desc table t;
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)
```
[ERROR] [conn.go:733] ["connection running loop panic"] [conn=3] [lastSQL="desc table t1"] [err="runtime error: slice bounds out of range [-1:]"] [stack="goroutine 616 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1(0x6aae200, 0xc000f65860, 0xc000137100)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/server/conn.go:731 +0xee\npanic(0x6296300, 0xc000d52680)\n\t/usr/local/Cellar/go/1.14.3/libexec/src/runtime/panic.go:969 +0x166\ngithub.com/pingcap/tidb/planner.extractSelectAndNormalizeDigest(0x6ad1f20, 0xc0010a3ec0, 0xc0005dcfbc, 0x4, 0xc000a1c5a0, 0xc000fc6640, 0xc0010c49c0, 0x6aca9c0, 0xc001038d00, 0xc000fc6640)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/planner/optimize.go:312 +0x84a\ngithub.com/pingcap/tidb/planner.getBindRecord(0x6b00b80, 0xc0009b3300, 0x6ad1f20, 0xc0010a3ec0, 0xc0010a3ec0, 0xc0010a3e01, 0x6af3740)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/planner/optimize.go:358 +0xb7\ngithub.com/pingcap/tidb/planner.Optimize(0x6aae200, 0xc0010c4840, 0x6b00b80, 0xc0009b3300, 0x6ac9580, 0xc0010a3ec0, 0x6af3740, 0xc000c7b830, 0x0, 0x0, ...)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/planner/optimize.go:134 +0x571\ngithub.com/pingcap/tidb/executor.(*Compiler).Compile(0xc000fc6dd8, 0x6aae200, 0xc0010c4840, 0x6ad1f20, 0xc0010a3ec0, 0x0, 0x0, 0x0)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/executor/compiler.go:62 +0x2fd\ngithub.com/pingcap/tidb/session.(*session).ExecuteStmt(0xc0009b3300, 0x6aae200, 0xc0010c4540, 0x6ad1f20, 0xc0010a3ec0, 0x0, 0x0, 0x0, 0x0)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/session/session.go:1379 +0x33a\ngithub.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt(0xc00101b050, 0x6aae200, 0xc0010c4540, 0x6ad1f20, 0xc0010a3ec0, 0xc0006a8210, 0x6aae200, 0xc0010c4540, 0x874ae58)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/server/driver_tidb.go:212 +0x68\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmt(0xc000137100, 0x6aae200, 0xc0010c4540, 0x6ad1f20, 0xc0010a3ec0, 0x874ae58, 0x0, 0x0, 0x1, 0x0, ...)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/server/conn.go:1561 +0x139\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc000137100, 0x6aae140, 0xc0010a3e40, 0xc000f5e431, 0xd, 0x0, 0x0)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/server/conn.go:1453 +0x49b\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch(0xc000137100, 0x6aae140, 0xc0010a3e40, 0xc000f5e430, 0xe, 0xd, 0x0, 0x0)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/server/conn.go:1027 +0x696\ngithub.com/pingcap/tidb/server.(*clientConn).Run(0xc000137100, 0x6aae200, 0xc000f65860)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/server/conn.go:792 +0x29c\ngithub.com/pingcap/tidb/server.(*Server).onConn(0xc0010c61a0, 0xc000137100)\n\t/Users/fzh/go/src/github.com/pingcap/tidb/server/server.go:461 +0xa97\ncreated by github.com/pingcap/tidb/server.(*Server).Run\n\t/Users/fzh/go/src/github.com/pingcap/tidb/server/server.go:364 +0x894\n"]

```
### 4. What is your TiDB version? (Required)
```
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: None
Edition: Community
Git Commit Hash: None
Git Branch: None
UTC Build Time: None
GoVersion: go1.14.3
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

```
<!-- Paste the output of SELECT tidb_version() -->

--- a/bindinfo/bind_test.go
+++ b/bindinfo/bind_test.go
@@ -2038,3 +2038,14 @@ func (s *testSuite) TestCaptureWithZeroSlowLogThreshold(c *C) {
 	c.Assert(len(rows), Equals, 1)
 	c.Assert(rows[0][0], Equals, "select * from test . t")
 }
+
+func (s *testSuite) TestExplainTableStmts(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	s.cleanBindingEnv(tk)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(id int, value decimal(5,2))")
+	tk.MustExec("table t")
+	tk.MustExec("explain table t")
+	tk.MustExec("desc table t")
+}

--- a/planner/core/preprocess.go
+++ b/planner/core/preprocess.go
@@ -249,6 +249,14 @@ func EraseLastSemicolon(stmt ast.StmtNode) {
 	}
 }
 
+// EraseLastSemicolonInSQL removes last semicolon of the SQL.
+func EraseLastSemicolonInSQL(sql string) string {
+	if len(sql) > 0 && sql[len(sql)-1] == ';' {
+		return sql[:len(sql)-1]
+	}
+	return sql
+}
+
 const (
 	// TypeInvalid for unexpected types.
 	TypeInvalid byte = iota

--- a/planner/optimize.go
+++ b/planner/optimize.go
@@ -131,7 +131,10 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 	if !ok {
 		return bestPlan, names, nil
 	}
-	bindRecord, scope := getBindRecord(sctx, stmtNode)
+	bindRecord, scope, err := getBindRecord(sctx, stmtNode)
+	if err != nil {
+		return nil, nil, err
+	}
 	if bindRecord == nil {
 		return bestPlan, names, nil
 	}
@@ -274,7 +277,7 @@ func optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 	return finalPlan, names, cost, err
 }
 
-func extractSelectAndNormalizeDigest(stmtNode ast.StmtNode, specifiledDB string) (ast.StmtNode, string, string) {
+func extractSelectAndNormalizeDigest(stmtNode ast.StmtNode, specifiledDB string) (ast.StmtNode, string, string, error) {
 	switch x := stmtNode.(type) {
 	case *ast.ExplainStmt:
 		// This function is only used to find bind record.
@@ -283,35 +286,19 @@ func extractSelectAndNormalizeDigest(stmtNode ast.StmtNode, specifiledDB string)
 		// The difference between them is whether len(x.Text()) is empty. They cannot be distinguished by stmt.restore.
 		// For these cases, we need return "" as normalize SQL and hash.
 		if len(x.Text()) == 0 {
-			return x.Stmt, "", ""
+			return x.Stmt, "", "", nil
 		}
 		switch x.Stmt.(type) {
 		case *ast.SelectStmt, *ast.DeleteStmt, *ast.UpdateStmt, *ast.InsertStmt:
-			plannercore.EraseLastSemicolon(x)
-			var normalizeExplainSQL string
+			var normalizeSQL string
 			if specifiledDB != "" {
-				normalizeExplainSQL = parser.Normalize(utilparser.RestoreWithDefaultDB(x, specifiledDB))
+				normalizeSQL = parser.Normalize(utilparser.RestoreWithDefaultDB(x.Stmt, specifiledDB))
 			} else {
-				normalizeExplainSQL = parser.Normalize(x.Text())
+				normalizeSQL = parser.Normalize(x.Text())
 			}
-			idx := int(0)
-			switch n := x.Stmt.(type) {
-			case *ast.SelectStmt:
-				idx = strings.Index(normalizeExplainSQL, "select")
-			case *ast.DeleteStmt:
-				idx = strings.Index(normalizeExplainSQL, "delete")
-			case *ast.UpdateStmt:
-				idx = strings.Index(normalizeExplainSQL, "update")
-			case *ast.InsertStmt:
-				if n.IsReplace {
-					idx = strings.Index(normalizeExplainSQL, "replace")
-				} else {
-					idx = strings.Index(normalizeExplainSQL, "insert")
-				}
-			}
-			normalizeSQL := normalizeExplainSQL[idx:]
+			normalizeSQL = plannercore.EraseLastSemicolonInSQL(normalizeSQL)
 			hash := parser.DigestNormalized(normalizeSQL)
-			return x.Stmt, normalizeSQL, hash
+			return x.Stmt, normalizeSQL, hash, nil
 		case *ast.SetOprStmt:
 			plannercore.EraseLastSemicolon(x)
 			var normalizeExplainSQL string
@@ -327,7 +314,7 @@ func extractSelectAndNormalizeDigest(stmtNode ast.StmtNode, specifiledDB string)
 			}
 			normalizeSQL := normalizeExplainSQL[idx:]
 			hash := parser.DigestNormalized(normalizeSQL)
-			return x.Stmt, normalizeSQL, hash
+			return x.Stmt, normalizeSQL, hash, nil
 		}
 	case *ast.SelectStmt, *ast.SetOprStmt, *ast.DeleteStmt, *ast.UpdateStmt, *ast.InsertStmt:
 		plannercore.EraseLastSemicolon(x)
@@ -337,7 +324,7 @@ func extractSelectAndNormalizeDigest(stmtNode ast.StmtNode, specifiledDB string)
 		// The difference between them is whether len(x.Text()) is empty. They cannot be distinguished by stmt.restore.
 		// For these cases, we need return "" as normalize SQL and hash.
 		if len(x.Text()) == 0 {
-			return x, "", ""
+			return x, "", "", nil
 		}
 		var normalizedSQL, hash string
 		if specifiledDB != "" {
@@ -345,34 +332,37 @@ func extractSelectAndNormalizeDigest(stmtNode ast.StmtNode, specifiledDB string)
 		} else {
 			normalizedSQL, hash = parser.NormalizeDigest(x.Text())
 		}
-		return x, normalizedSQL, hash
+		return x, normalizedSQL, hash, nil
 	}
-	return nil, "", ""
+	return nil, "", "", nil
 }
 
-func getBindRecord(ctx sessionctx.Context, stmt ast.StmtNode) (*bindinfo.BindRecord, string) {
+func getBindRecord(ctx sessionctx.Context, stmt ast.StmtNode) (*bindinfo.BindRecord, string, error) {
 	// When the domain is initializing, the bind will be nil.
 	if ctx.Value(bindinfo.SessionBindInfoKeyType) == nil {
-		return nil, ""
+		return nil, "", nil
 	}
-	stmtNode, normalizedSQL, hash := extractSelectAndNormalizeDigest(stmt, ctx.GetSessionVars().CurrentDB)
+	stmtNode, normalizedSQL, hash, err := extractSelectAndNormalizeDigest(stmt, ctx.GetSessionVars().CurrentDB)
+	if err != nil {
+		return nil, "", err
+	}
 	if stmtNode == nil {
-		return nil, ""
+		return nil, "", nil
 	}
 	sessionHandle := ctx.Value(bindinfo.SessionBindInfoKeyType).(*bindinfo.SessionHandle)
 	bindRecord := sessionHandle.GetBindRecord(normalizedSQL, "")
 	if bindRecord != nil {
 		if bindRecord.HasUsingBinding() {
-			return bindRecord, metrics.ScopeSession
+			return bindRecord, metrics.ScopeSession, nil
 		}
-		return nil, ""
+		return nil, "", nil
 	}
 	globalHandle := domain.GetDomain(ctx).BindHandle()
 	if globalHandle == nil {
-		return nil, ""
+		return nil, "", nil
 	}
 	bindRecord = globalHandle.GetBindRecord(hash, normalizedSQL, "")
-	return bindRecord, metrics.ScopeGlobal
+	return bindRecord, metrics.ScopeGlobal, nil
 }
 
 func handleInvalidBindRecord(ctx context.Context, sctx sessionctx.Context, level string, bindRecord bindinfo.BindRecord) {

--- a/planner/optimize.go
+++ b/planner/optimize.go
@@ -343,11 +343,8 @@ func getBindRecord(ctx sessionctx.Context, stmt ast.StmtNode) (*bindinfo.BindRec
 		return nil, "", nil
 	}
 	stmtNode, normalizedSQL, hash, err := extractSelectAndNormalizeDigest(stmt, ctx.GetSessionVars().CurrentDB)
-	if err != nil {
+	if err != nil || stmtNode == nil {
 		return nil, "", err
-	}
-	if stmtNode == nil {
-		return nil, "", nil
 	}
 	sessionHandle := ctx.Value(bindinfo.SessionBindInfoKeyType).(*bindinfo.SessionHandle)
 	bindRecord := sessionHandle.GetBindRecord(normalizedSQL, "")

